File: packages/pvm-assemblyscript/assembly/instructions/shifts.ts
import {
  OPCODE_NEG_ADD_IMM_32,
  OPCODE_SHAR_R_IMM_32,
  OPCODE_SHLO_L_IMM_32,
  OPCODE_SHLO_R_IMM_32,
} from '../config'
import { InstructionContext, InstructionResult } from '../types'
import { BaseInstruction } from './base'

/**
 * SHLO_L_IMM_32 instruction (opcode 0x8A)
 * Gray Paper formula: reg'_A = sext{4}{(reg_B · 2^(immed_X mod 32)) mod 2^32}
 */
export class SHLO_L_IMM_32Instruction extends BaseInstruction {
  opcode: i32 = OPCODE_SHLO_L_IMM_32
  name: string = 'SHLO_L_IMM_32'

  execute(context: InstructionContext): InstructionResult {
    // Gray Paper formula: reg'_A = sext{4}{(reg_B · 2^(immed_X mod 32)) mod 2^32}
    const parseResult = this.parseTwoRegistersAndImmediate(context.operands, context.fskip)
    const registerA = parseResult.registerA
    const registerB = parseResult.registerB
    const immediateX = parseResult.immediateX
    const registerValue = this.getRegisterValueAs32(context.registers, registerB)

    // Ensure shift amount is within 32-bit range
    const immediate32 = u64(immediateX) & u64(0xffffffff) // Convert to 32-bit number
    const shiftAmount = immediate32 % u64(32)
    const shifted = u64(registerValue) << shiftAmount
    const result = shifted & u64(0xffffffff) // Mask to 32 bits

    this.setRegisterValueWith32BitResult(context.registers, registerA, result)

    return new InstructionResult(-1)
  }
}

/**
 * SHLO_R_IMM_32 instruction (opcode 0x8B)
 * Gray Paper formula: reg'_A = sext{4}{floor(reg_B mod 2^32 / 2^(immed_X mod 32))}
 */
export class SHLO_R_IMM_32Instruction extends BaseInstruction {
  opcode: i32 = OPCODE_SHLO_R_IMM_32
  name: string = 'SHLO_R_IMM_32'

  execute(context: InstructionContext): InstructionResult {
    // Gray Paper formula: reg'_A = sext{4}{floor(reg_B mod 2^32 / 2^(immed_X mod 32))}
    const parseResult = this.parseTwoRegistersAndImmediate(context.operands, context.fskip)
    const registerA = parseResult.registerA
    const registerB = parseResult.registerB
    const immediateX = parseResult.immediateX
    const registerValue = this.getRegisterValueAs32(context.registers, registerB)

    // Ensure shift amount is within 32-bit range
    const immediate32 = u64(immediateX) & u64(0xffffffff) // Convert to 32-bit number
    const shiftAmount = immediate32 % u64(32)
    const shifted = u64(registerValue) >> shiftAmount // Right shift
    const result = shifted & u64(0xffffffff) // Mask to 32 bits

    this.setRegisterValueWith32BitResult(context.registers, registerA, result)

    return new InstructionResult(-1)
  }
}

/**
 * SHAR_R_IMM_32 instruction (opcode 0x8C)
 * Gray Paper formula: reg'_A = unsigned{floor(signed_4(reg_B mod 2^32) / 2^(immed_X mod 32))}
 */
export class SHAR_R_IMM_32Instruction extends BaseInstruction {
  opcode: i32 = OPCODE_SHAR_R_IMM_32
  name: string = 'SHAR_R_IMM_32'

  execute(context: InstructionContext): InstructionResult {
    // Gray Paper formula: reg'_A = unsigned{floor(signed_4(reg_B mod 2^32) / 2^(immed_X mod 32))}
    const parseResult = this.parseTwoRegistersAndImmediate(context.operands, context.fskip)
    const registerA = parseResult.registerA
    const registerB = parseResult.registerB
    const immediateX = parseResult.immediateX
    const registerValue = this.getRegisterValueAs32(context.registers, registerB)

    // Arithmetic shift (signed)
    const immediate32 = u64(immediateX) & u64(0xffffffff) // Convert to 32-bit number
    const shiftAmount = immediate32 % u64(32)

    // Convert to signed 32-bit, perform arithmetic right shift, then back to unsigned
    const signedValue_u64 = this.signExtend(registerValue, 4) // Convert to signed 32-bit (returns u64)
    const signedValue_i64 = i64(signedValue_u64) // Cast to i64 for arithmetic right shift
    const shifted = signedValue_i64 >> i64(shiftAmount) // Arithmetic right shift
    const result = u64(shifted) & u64(0xffffffff) // Convert back to unsigned 32-bit

    this.setRegisterValueWith32BitResult(context.registers, registerA, result)

    return new InstructionResult(-1)
  }
}

/**
 * NEG_ADD_IMM_32 instruction (opcode 0x8D)
 * Gray Paper formula: reg'_A = sext{4}{(immed_X + 2^32 - reg_B) mod 2^32}
 */
export class NEG_ADD_IMM_32Instruction extends BaseInstruction {
  opcode: i32 = OPCODE_NEG_ADD_IMM_32
  name: string = 'NEG_ADD_IMM_32'

  execute(context: InstructionContext): InstructionResult {
    // Use the standard two-register-and-immediate format
    const parseResult = this.parseTwoRegistersAndImmediate(context.operands, context.fskip)
    const registerA = parseResult.registerA
    const registerB = parseResult.registerB
    const immediateX = parseResult.immediateX
    const registerValue = this.getRegisterValueAs32(context.registers, registerB)

    // Gray Paper formula: reg'_A = (immed_X + 2^32 - reg_B) mod 2^32
    const immediate32 = u64(immediateX) & u64(0xffffffff)
    const result = (immediate32 + u64(0x100000000) - u64(registerValue)) & u64(0xffffffff)

    this.setRegisterValueWith32BitResult(context.registers, registerA, result)

    return new InstructionResult(-1)
  }
}

File: packages/pvm-rust/src/instructions/shifts.rs
//! Shift by immediate (32-bit): SHLO_L, SHLO_R, SHAR_R, NEG_ADD. Gray Paper §7.4.9.

use crate::config::{
    OPCODE_NEG_ADD_IMM_32, OPCODE_SHAR_R_IMM_32, OPCODE_SHLO_L_IMM_32, OPCODE_SHLO_R_IMM_32,
};
use crate::instructions::base::{
    arithmetic_shift_right_32, parse_two_registers_and_immediate, sign_extend, InstructionHandler,
};
use crate::types::{InstructionContext, InstructionResult};

fn get_register(registers: &[u64; 13], index: u8) -> u64 {
    registers.get(index as usize).copied().unwrap_or(0)
}

fn set_register_32(registers: &mut [u64; 13], index: u8, value: u64) {
    let extended = sign_extend(value & 0xffff_ffff, 4);
    if (index as usize) < 13 {
        registers[index as usize] = extended;
    }
}

// SHLO_L_IMM_32: reg'_A = sext{4}{(reg_B << (immed_X mod 32)) mod 2^32}
pub struct ShloLImm32Instruction;
impl ShloLImm32Instruction {
    #[must_use]
    pub const fn new() -> Self {
        Self
    }
}
impl InstructionHandler for ShloLImm32Instruction {
    fn opcode(&self) -> i32 {
        i32::from(OPCODE_SHLO_L_IMM_32)
    }
    fn name(&self) -> &'static str {
        "SHLO_L_IMM_32"
    }
    fn execute(&self, context: &mut InstructionContext<'_>) -> InstructionResult {
        let p = parse_two_registers_and_immediate(context.operands, context.fskip);
        let reg_b = get_register(context.registers, p.register_b) & 0xffff_ffff;
        let shift = (p.immediate_x as u64) & 0xffff_ffff;
        let shift = shift % 32;
        let result = (reg_b << shift) & 0xffff_ffff;
        set_register_32(context.registers, p.register_a, result);
        InstructionResult::new(InstructionResult::CONTINUE, 0)
    }
}

// SHLO_R_IMM_32: reg'_A = sext{4}{floor(reg_B mod 2^32 / 2^(immed_X mod 32))}
pub struct ShloRImm32Instruction;
impl ShloRImm32Instruction {
    #[must_use]
    pub const fn new() -> Self {
        Self
    }
}
impl InstructionHandler for ShloRImm32Instruction {
    fn opcode(&self) -> i32 {
        i32::from(OPCODE_SHLO_R_IMM_32)
    }
    fn name(&self) -> &'static str {
        "SHLO_R_IMM_32"
    }
    fn execute(&self, context: &mut InstructionContext<'_>) -> InstructionResult {
        let p = parse_two_registers_and_immediate(context.operands, context.fskip);
        let reg_b = get_register(context.registers, p.register_b) & 0xffff_ffff;
        let shift = (p.immediate_x as u64) & 0xffff_ffff;
        let shift = (shift % 32) as u32;
        let result = if shift == 0 {
            reg_b
        } else {
            reg_b >> shift
        };
        set_register_32(context.registers, p.register_a, result);
        InstructionResult::new(InstructionResult::CONTINUE, 0)
    }
}

// SHAR_R_IMM_32: reg'_A = unsigned{floor(signed_4(reg_B mod 2^32) / 2^(immed_X mod 32))}
pub struct SharRImm32Instruction;
impl SharRImm32Instruction {
    #[must_use]
    pub const fn new() -> Self {
        Self
    }
}
impl InstructionHandler for SharRImm32Instruction {
    fn opcode(&self) -> i32 {
        i32::from(OPCODE_SHAR_R_IMM_32)
    }
    fn name(&self) -> &'static str {
        "SHAR_R_IMM_32"
    }
    fn execute(&self, context: &mut InstructionContext<'_>) -> InstructionResult {
        let p = parse_two_registers_and_immediate(context.operands, context.fskip);
        let reg_b = get_register(context.registers, p.register_b) & 0xffff_ffff;
        let shift = (p.immediate_x as u64) & 0xffff_ffff;
        let shift = (shift % 32) as u32;
        let result = arithmetic_shift_right_32(reg_b, shift);
        set_register_32(context.registers, p.register_a, result);
        InstructionResult::new(InstructionResult::CONTINUE, 0)
    }
}

// NEG_ADD_IMM_32: reg'_A = sext{4}{(immed_X + 2^32 - reg_B) mod 2^32}
pub struct NegAddImm32Instruction;
impl NegAddImm32Instruction {
    #[must_use]
    pub const fn new() -> Self {
        Self
    }
}
impl InstructionHandler for NegAddImm32Instruction {
    fn opcode(&self) -> i32 {
        i32::from(OPCODE_NEG_ADD_IMM_32)
    }
    fn name(&self) -> &'static str {
        "NEG_ADD_IMM_32"
    }
    fn execute(&self, context: &mut InstructionContext<'_>) -> InstructionResult {
        let p = parse_two_registers_and_immediate(context.operands, context.fskip);
        let reg_b = get_register(context.registers, p.register_b) & 0xffff_ffff;
        let imm = (p.immediate_x as u64) & 0xffff_ffff;
        let result = (imm.wrapping_add(0x1_0000_0000).wrapping_sub(reg_b)) & 0xffff_ffff;
        set_register_32(context.registers, p.register_a, result);
        InstructionResult::new(InstructionResult::CONTINUE, 0)
    }
}
